flux-accounting: add a local job-archive

cmoussa1 opened the pull request:
#### Background

As discussed in https://github.com/flux-framework/flux-core/issues/3136 and in #353, it may be advantageous for flux accounting to do its own job record archival. With archival moved to flux accounting, the job-archive module in flux-core could be retired and the flux accounting project can evolve the schema as needed to fit its requirements.

---

This PR looks to begin to add a job archive local to flux-accounting. It creates two new tables in the flux-accounting DB:

- `jobs`: a table that stores job records used for calculating job usage (and subsequently fair share)
- `last_seen_job_table`: a table that stores a timestamp of the last inactive job seen

It also adds a new Python script that is designed to be run as a cron job (and in conjunction with the other Python scripts and commands in this repository that run periodically). This new Python script, `flux-account-fetch-job-records.py`, will fetch inactive jobs using Flux's `job-info` and `job-list` interfaces. It will create job-record objects and insert them into the newly created `jobs` table. The benefit here is when it comes time for flux-accounting to re-calculate job-usage and fair share values for user/banks during a periodic update, it can just look locally in its own DB instead of having to connect to flux-core's job-archive.

Some minor cleanup of the `update-usage` and `view-job-records` commands (as well as some of the unit tests for the portion of flux-accounting that previously dealt with the flux-core job-archive) follow the first couple commits that add the new tables and Python script. No major changes were required to the tests that previously dealt with the flux-core job-archive (specifically `t1011-job-archive-interface.t`). The big change there was to add a call to the new Python script and point the `view-job-records` and `update-usage` commands to the flux-accounting DB.

Note that this new Python script does not yet have a "purge" function just yet to remove old jobs that are no longer considered by job usage or fair share calculations. For the sake of keeping this PR as narrow in scope as possible, I figured I would open a PR that follows this one which looks to add purge capabilities. 

---

You'll also notice that right now there is a commit in this PR that removes one of the sharness tests (`t1010-update-usage.t`) from `t/Makefile.am`, essentially ignoring it when running `make check`. The reason for this is `t1010-job-usage.t` creates a fake job-archive DB to test the `update-usage` command for flux-accounting. This test was created before additional tests were added to flux-accounting that _actually_ loaded flux-core's job-archive. But now that flux-accounting has implemented its own job-archive, I don't think this sharness test file is necessarily needed, especially because the next sharness test file, `t1011-job-archive-interface.db`, actually tests fetching job records and updating usage using (now) flux-accounting's job-archive. So I'm not totally sure what the best approach here is. Should I remove the test file completely from the repository and re-number the following sharness tests? Is leaving it out of `t/Makefile.am` fine? Let me know.

cmoussa1 commented:
OK, I think I'll mark this is as ready for some initial review whenever convenient. 😄

garlick commented:
Awesome @cmoussa1 !

> Should I remove the test file completely from the repository and re-number the following sharness tests? Is leaving it out of t/Makefile.am fine? Let me know.

IMHO if the test is no longer needed, just remove it.  No need to renumber the tests.  Just leave a hole :-)

I'll plan to make a review pass later today and test this on my small cluster FWIW.  I think it would be good for @chu11 and @ryanday36 to also have a look, keeping in mind this would be intended to _replace_ job-archive. 

Maybe it would make sense to create a PR for the flux accounting guide that remains pending until this is included in a new tag?  The stuff in there about configuring job-archive could be dropped and a suggested crontab line to run the new script periodically could be added to the example there (which I would cut and paste for my testing).

cmoussa1 commented:
Thanks @garlick!

> IMHO if the test is no longer needed, just remove it. No need to renumber the tests. Just leave a hole :-)

That sounds good. I'll edit the commit to just remove it. Thanks!

> I'll plan to make a review pass later today and test this on my small cluster FWIW. I think it would be good for @chu11 and @ryanday36 to also have a look, keeping in mind this would be intended to replace job-archive.

Yes, this sounds great. Note that I leave for vacation in a couple hours so I'll probably be delayed in a response to your review comments. 😅 

> Maybe it would make sense to create a PR for the flux accounting guide that remains pending until this is included in a new tag? The stuff in there about configuring job-archive could be dropped and a suggested crontab line to run the new script periodically could be added to the example there (which I would cut and paste for my testing).

I agree! I'll create an issue for it so I don't forget and try to open one soon. 👍

cmoussa1 commented:
Note: I've updated this PR to use the new flux-core job-info convenience function that was added in flux-framework/flux-core#5265

cmoussa1 commented:
just rebased to catch up after #367

garlick commented:
Poking at this a bit (finally!)

To get it going I rebased on current master in a private repo and build a deb, then installed that on my test cluster rank 0.  Then I ran 
```
$ sudo -u flux flux account-update-db
checking for new tables...
new table found: jobs
new table found: last_seen_job_table
checking for new columns to add in tables...
```
So far so good.  For fun I poked at the db and dumped the definition of the `jobs` table:
```
$ sqlite3 FluxAccounting.db
SQLite version 3.34.1 2021-01-20 14:10:07
Enter ".help" for usage hints.
sqlite> .schema jobs
CREATE TABLE IF NOT EXISTS "jobs" (id integer  NOT NULL, userid integer  NOT NULL, t_submit real  NOT NULL, t_run real  NOT NULL, t_inactive real  NOT NULL, ranks tinytext  NOT NULL, R tinytext  NOT NULL, jobspec tinytext  NOT NULL, PRIMARY KEY (id));
```
And noted a few differences from the job-archive `jobs` table (no problem - just noting!):
```
$ sqlite3 job-archive.sqlite
SQLite version 3.34.1 2021-01-20 14:10:07
Enter ".help" for usage hints.
sqlite> .schema jobs
CREATE TABLE jobs(  id CHAR(16) PRIMARY KEY,  userid INT,  ranks TEXT,  t_submit REAL,  t_run REAL,  t_cleanup REAL,  t_inactive REAL,  eventlog TEXT,  jobspec TEXT,  R TEXT);
```
Then I tried manually populating the new table with the `fetch-job-records` script but that didn't go so well:
```
$ flux account-fetch-job-records
Traceback (most recent call last):
  File "/usr/lib/aarch64-linux-gnu/flux/cmd/flux-account-fetch-job-records.py", line 171, in <module>
    main()
  File "/usr/lib/aarch64-linux-gnu/flux/cmd/flux-account-fetch-job-records.py", line 156, in main
    timestamp = get_last_job_ts(conn)
  File "/usr/lib/aarch64-linux-gnu/flux/cmd/flux-account-fetch-job-records.py", line 65, in get_last_job_ts
    return float(row[0])
TypeError: 'NoneType' object is not subscriptable
```
Probably unrelated but I also noticed that for me I got a couple of errors when running `make check` - were these supposed to have been removed?
```
ERROR: test/test_job_archive_interface.py - missing test plan
ERROR: test/test_job_archive_interface.py - exited with status 1
```
I thought maybe I'd pause here and ask you to rebase this PR on current master and make sure everything still works for you.

garlick commented:
Two notes on the schema:
- I don't think TINYTEXT is a sqlite type - it is accepted but converted to TEXT
- A sqlite INTEGER is a 64 bit signed value, but job IDs (in integer form) use the full 64 bit unsigned range.  Maybe that is why @chu11 used CHAR(16) as the type for job IDs?  Although maybe INTEGER would work and be more efficient as long as the sign conversion is carefully handled (a test would be good).

Also the subscript in CHAR(16) is just another alias for TEXT in sqlite (and subscripts are ignored)

https://www.sqlite.org/draft/datatype3.html

cmoussa1 commented:
Thanks for beginning to take a look at this @garlick! I spent some time this morning looking at your feedback; sorry that it didn't work out so gracefully on your test cluster this first time around!

> And noted a few differences from the job-archive jobs table (no problem - just noting!):

Thanks for providing this comparison. Yeah, I noticed that the types for some of the data that flux-accounting is interested in was different in flux-accounting's job-archive and flux-core's, so I went ahead and changed them to match the data types of flux-core's job-archive. They should match now.

> Probably unrelated but I also noticed that for me I got a couple of errors when running make check - were these supposed to have been removed?
> ```
> ERROR: test/test_job_archive_interface.py - missing test plan
> ERROR: test/test_job_archive_interface.py - exited with status 1
> ```

That is odd - that set of tests is supposed to be there and run. I have never run `make deb` before (I am unsure how), however, so I wonder if there is something weird going on with this unit test during that build process? FWIW, this unit test runs in CI.

> Edit: I guess one other question is any reason not to use the same technique as @chu11 for getting the last job ID based on ? > the inactive time stamp? E.g.
>
> ```
> SELECT MAX(t_inactive) FROM jobs
> ```
> It might be a bit simpler than managing the last_seen_job_table (I'm assuming forgetting to add it to the schema update > script is what caused the error above)

Ah, that would be because I forgot you could query for MIN and MAX values, :-) Thanks for pointing this out. I went ahead and force-pushed a change to this script to run `SELECT MAX(t_inactive) FROM jobs` when looking for the most recent inactive job in the jobs table. If the `jobs` table is empty, the variable that holds the timestamp of the most recent inactive job should just be `0.0` (indicating it should look for _all_ inactive jobs).

I've also rebased on current master, so it should be caught up after #369 now.

garlick commented:
Thanks @cmoussa - i've updated to your branch, and I'm still getting those two test failures.  The logs say
```
Traceback (most recent call last):
  File "/nfshome/garlick/proj/flux-accounting/src/bindings/python/fluxacct/accounting/./test/test_job_archive_interface.py", line 20, in <module>
    from fluxacct.accounting import job_archive_interface as jobs
  File "/nfshome/garlick/proj/flux-accounting/src/bindings/python/fluxacct/accounting/job_archive_interface.py", line 18, in <module>
    from flux.resource import ResourceSet
ModuleNotFoundError: No module named 'flux'
ERROR: test/test_job_archive_interface.py - missing test plan
ERROR: test/test_job_archive_interface.py - exited with status 1
```
I'll see if I can find out why that's failing on my system.

Note: the `make deb` just builds a deb package for the project (like an RPM).  It shouldn't affect the tests in any way.

garlick commented:
Those tests are failing for me on master also so I guess it's not related to this PR.

cmoussa1 commented:
Thanks for letting me know @garlick!

```
Traceback (most recent call last):
  File "/nfshome/garlick/proj/flux-accounting/src/bindings/python/fluxacct/accounting/./test/test_job_archive_interface.py", line 20, in <module>
    from fluxacct.accounting import job_archive_interface as jobs
  File "/nfshome/garlick/proj/flux-accounting/src/bindings/python/fluxacct/accounting/job_archive_interface.py", line 18, in <module>
    from flux.resource import ResourceSet
ModuleNotFoundError: No module named 'flux'
ERROR: test/test_job_archive_interface.py - missing test plan
ERROR: test/test_job_archive_interface.py - exited with status 1
```

It looks like from the description of this error that flux-accounting is having trouble `import`ing a Python module from flux-core: `flux.resource` is a Python module that allows flux-accounting to fetch `nnodes` from a job's R.

garlick commented:
One thing we might need here is a script to import data from an existing job-archive database.

cmoussa1 commented:
Ah, like something that can parse a _flux-core_ job-archive and import it to a _flux-accounting_ job-archive?

garlick commented:
Yes - I'm guessing we will not want to lose the historical data we've already collected...

cmoussa1 commented:
That makes sense, and I think it could be more than reasonable to develop a Python script or something that can fetch and parse records from a flux-core job-archive and import it to a flux-accounting job-archive. Are you thinking that the script could parse just what it might need for flux-accounting stuff? i.e just fetch the data in the schema proposed in this PR so far?

garlick commented:
The schemas are pretty close so I guess I was thinking that instead of creating an empty database when one does not exist, we could optionally copy from an existing one?  Just one time only.

Then we can eventually remove the old database (and the core module that populates it).

cmoussa1 commented:
> The schemas are pretty close so I guess I was thinking that instead of creating an empty database when one does not exist, we could optionally copy from an existing one? Just one time only.

OK, I think that sounds reasonable; I'll try to take a crack at this and edit this PR with the new feature.

cmoussa1 commented:
Ah, I'm sorry, @garlick, I started working on this this morning and now I think I might not have understood you fully. 🤦 I think I might need more explanation on what you mean by copying from an old database.

What is the difference between fetching jobs from a flux-core job-archive DB and populating those records into a flux-accounting job-archive table (what is currently proposed here) and copying/importing data from an existing flux-core job-archive database?

garlick commented:
> What is the difference between fetching jobs from a flux-core job-archive DB and populating those records into a flux-accounting job-archive table (what is currently proposed here) and copying/importing data from an existing flux-core job-archive database?

Apologies, I think I failed to get my idea across!   I just meant that we provide a migration tool to copy data from the old database to the new database.  Then we could remove the old database without losing  _data_.  Does that make sense?

cmoussa1 commented:
> I just meant to propose that we provide a migration tool to copy data from the old database to the new database. Then we could remove the old database without losing _data_. Does that make sense?

Ohhh, so something that could connect to the old DB and old `jobs` table, grab all existing data, and then `INSERT` it into the new `jobs` table in the flux-accounting database?

garlick commented:
> Ohhh, so something that could connect to the old DB and old jobs table, grab all existing data, and then INSERT it into the new jobs table in the flux-accounting database?

Exactly :smile:  Sorry I wasn't clear.

cmoussa1 commented:
Haha no problem! Thanks for that clarification. Yes, I agree and think that would be a useful option for this script to have, maybe as an optional arg? I'll work on this!

garlick commented:
Sure or it could just include logic like
```
if JOBS table does not exist
    create JOBS table
    if job-archive.sqlite exists
        copy data from job-archive.sqlite to JOBS table
    endif
endif
```

cmoussa1 commented:
@garlick BTW, this is the PR that is looking to add a flux-accounting-specific job-archive that we were talking about yesterday. I _believe_ I tried to address your feedback about enabling a copy of an old job-archive's records to flux-accounting, but feel free to let me know if I'm on base here and should refactor. :-)

chu11 commented:
slowly starting to review this, and just wondering what is the plan to transition from the old `job-archive`?  perhaps there needs to be a script to do a "one time copy" of what you need from `job-archive`? (apologies if this is discussed above and I missed it).

cmoussa1 commented:
was just typing up a response! Sorry that I didn't make that clearer @chu11 - I should update the PR description

cmoussa1 commented on src/cmd/flux-account-fetch-job-records.py:
```diff
@@ -0,0 +1,195 @@
+    return job_records
+
+
+# insert newly seen jobs into the "jobs" table in the flux-accounting DB
+def insert_jobs_in_db(conn, job_records):
+    cur = conn.cursor()
+
+    for single_job in job_records:
+        cur.execute(
+            """
+            INSERT OR IGNORE INTO jobs
+            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
+            """,
+            (
+                single_job["id"],
+                single_job["userid"],
+                single_job["t_submit"],
+                single_job["t_run"],
+                single_job["t_inactive"],
+                single_job["ranks"],
+                single_job["R"],
+                single_job["jobspec"],
+            ),
+        )
+
+    conn.commit()
```
That's a good suggestion, thanks - you're right, I should probably check to see if a job actually ran. I'll go ahead and add a check as well as a test to submit a job that never runs and ensures that the job record doesn't show up under the user and that their job usage value doesn't change.

cmoussa1 commented on src/cmd/flux-account-fetch-job-records.py:
```diff
@@ -0,0 +1,195 @@
+def est_sqlite_conn(path):
+    if not os.path.isfile(path):
+        print(f"Database file does not exist: {path}", file=sys.stderr)
+        sys.exit(1)
```
thanks - just pushed up a change to fix the comment

cmoussa1 commented:
Thanks @chu11! I have an open PR #446 that looks to move the accounting guide to this repository; I should add some additional instructions to that guide regarding the change to fetching job records. I'll update that PR.

@ryanday36 - I'll try to keep you posted when a new version of flux-accounting is created so that we can coordinate in case there are hiccups.

I'll set MWP here
